bouffalo-hal/src/sdio.rs

```diff
@@ -3510,20 +3533,470 @@ impl TUNINGConfiguration {
     // TODO
 }
 
+/// SDH transfer flag.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum SDHTransFlag {
+    None = 0x00000000,
+    EnDma = 0x00000001,              // Enable DMA.
+    EnBlkCount = 0x00000002,         // Enable block count.
+    EnAutoCmd12 = 0x00000004,        // Enable auto CMD12.
+    EnAutoCmd23 = 0x00000008,        // Enable auto CMD23.
+    ReadData = 0x00000010,           // Enable read data.
+    MultiBlk = 0x00000020,           // Enable multi-block data operation.
+    Resp136Bits = 0x00010000,        // Response is 136 bits length.
+    Resp48Bits = 0x00020000,         // Response is 48 bits length.
+    Resp48BitsWithBusy = 0x00030000, // Response is 48 bits length with busy status.
+    EnCrcCheck = 0x00080000,         // Enable CRC check.
+    EnIndexCheck = 0x00100000,       // Enable index check.
+    DataPresent = 0x00200000,        // Data present.
+    Suspend = 0x00400000,            // Suspend command.
+    Resume = 0x00800000,             // Resume command.
+    Abort = 0x00C00000,              // Abort command.
+}
+
+/// SDH response type.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum SDHResp {
+    None,
+    R1,
+    R5,
+    R6,
+    R7,
+    R1B,
+    R5B,
+    R2,
+    R3,
+    R4,
+}
+
+/// Sleep for n milliseconds.
+pub fn sleep_ms(n: u32) {
+    for _ in 0..n * 125 {
+        unsafe { asm!("nop") }
+    }
+}
+
+/// Managed Secure Digital Host Controller peripheral.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct Sdh<SDH, PADS, const I: usize> {
+    sdh: SDH,
+    pads: PADS,
+    block_count: u32,
+}
+
+impl<SDH: Deref<Target = RegisterBlock>, PADS, const I: usize> Sdh<SDH, PADS, I> {
+    /// Create a new instance of the SDH peripheral.
+    #[inline]
+    pub fn new(sdh: SDH, pads: PADS) -> Self
+    where
+        PADS: Pads<I>,
+    {
+        let block_count = 0;
+        Self {
+            sdh,
+            pads,
+            block_count,
+        }
+    }
+
+    /// Initialize the SDH peripheral (enable debug to print card info).
+    #[inline]
+    pub fn init<GLB, W: Write>(&mut self, w: &mut W, glb: &GLB, debug: bool)
+    where
+        GLB: Deref<Target = glb::v2::RegisterBlock>,
+    {
+        // SDH_RESET.
+        let tmp_val = self.sdh.software_reset.read().reset_all();
+        unsafe {
+            self.sdh.software_reset.write(tmp_val);
+        }
+        while !self.sdh.software_reset.read().is_reset_all_finished() {
+            // Wait for software reset finished.
+            unsafe {
+                asm!("nop");
+            }
+        }
```

Comment:
```suggestion
        // Wait for software reset finished.
        while !self.sdh.software_reset.read().is_reset_all_finished() {
            core::hint::spin_loop()
        }
```

Use `core::hint::spin_loop()` for busy-wait loops. This function provides appropriate hint instructions for each architecture to temporarily reduce the instruction retirement rate, improving energy efficiency.